private/numerics/functions/elliptic.rkt

```diff
@@ -0,0 +1,241 @@
+           [xp (* .25 (+ x λₐ))]
+           [yp (* 1/4 (+ y λₐ))]
+           [zp (* .25 (+ z λₐ))]
+           [μ (* .2 (+ xp yp (* 3. zp)))]
+           [Δx (/ (- μ xp) μ)]
+           [Δy (/ (- μ yp) μ)]
+           [Δz (/ (- μ zp) μ)])
+      (if (> (max (abs Δx) (abs Δy) (abs Δz)) ε)
+          (loop xp yp zp sump facp)
+          (let* ([ea (* Δx Δy)]
+                 [eb (* Δz Δz)]
+                 [ec (- ea eb)]
+                 [ed (- ea (* 6. eb))]
+                 [ee (+ ed ec ec)])
+            (+ (* 3 sump)
+               (/ (* facp
+                     (+ (+ 1.
+                           (* ed
+                              (+ C₁ (* C₅ ed) (* C₆ Δz ee))))
+                        (* Δz (+ (* C₂ ee) (* Δz (+ (* C₃ ec) (* Δz C₄ ea)))))))
+                  (* μ (sqrt μ)))))))))
+
+(define Carlson-elliptic₂ Rd)
+(define Carlson-elliptic-2 Rd)
+
+(define (square x) (* x x))
+
+(define (elliptic-integral-F φ k)
+  (define sinφ (sin φ))
+  (* sinφ
+     (Carlson-elliptic₁ (square (cos φ))
+                        (* (- 1. (* sinφ k))
+                           (+ 1. (* sinφ k)))
+                        1.)))
+
+(define π pi)
+(define π/2 (/ π 2))
+(define (complete-elliptic-integral-K k)
+  (elliptic-integral-F π/2 k))
+
+(define (elliptic-integral-E φ k)
+  (define sinφ (sin φ))
+  (define cos²φ (square (cos φ)))
+  (define q (* (- 1. (* sinφ k))
+               (+ 1. (* sinφ k))))
+  (* sinφ
+     (- (Carlson-elliptic₁ cos²φ q 1.)
+        (* (square (* sinφ k))
+           (/ (Carlson-elliptic₂ cos²φ q 1.) 3.)))))
+
+(define (complete-elliptic-integral-E k)
+  (elliptic-integral-E π/2 k))
+
+;;; older definition of the complete elliptic integrals
+;;; probably from A&Stegun
+
+(define (elliptic-integrals k continue)
+  (if (= k 1)
+      (continue +inf.0 1.)
+      (let loop ([a 1.0]
+                 [b (sqrt (- 1.0 (square k)))]
+                 [c k]
+                 [d 0.0]
+                 [powers-2 1.0])
+        (if (< (abs c) machine-ε)
+            (let ([first-elliptic-integral (/ π/2 a)])
+              (continue first-elliptic-integral
+                        (* first-elliptic-integral
+                           (- 1.0 (/ d 2.0)))))
+            (loop (/ (+ a b) 2.0)
+                  (sqrt (* a b))
+                  (/ (- a b) 2.0)
+                  (+ d (* (square c)
+                          powers-2))
+                  (( powers-2 2.0)))))))
+
+;; K
+(define (first-elliptic-integral k)
+  (elliptic-integrals k (λ (K E) K)))
+
+;; E
+(define (second-elliptic-integral k)
+  (elliptic-integrals k (λ (K E) E)))
+
+(define (first-elliptic-integral&derivative k cont)
+  (if (= k 0.0)
+      (cons π/2 0.0)
+      (elliptic-integrals
+       k
+       (λ (Kk Ek)
+         (cont Kk
+               (/ (- (/ Ek (- 1 (square k)))
+                     Kk)
+                  k))))))
+
+;; TODO: refactor this, but write tests first to make sure we didnt
+;; break anything.
+(define Jacobi-elliptic-functions
```

Comment:
Yea, I tried my luck at refactoring a bit but just could get it to return the same outputs. There's to many nested lets, named let loops in the middle, and I don't understand the variable naming scheme. I think this was probably translated verbatim from some textbook with a FORTRAN or ALGOL version of this algorithm. I'll try to hunt down the reference and see if I can reimplement from there.